### PR TITLE
Return NaN for out of bound conversions in FITS WCS

### PR DESCRIPTION
**Pull request by astrofrog**
I updated the APE 14 base class to say 'should' instead of 'can' because I think it's important we have consistent behavior.

### CODE REVIEW

**Comment by pllim** [comment]
I don't think this should be backported.

**Comment by astrofrog** [comment]
@pllim - we don't need to, the APE says 'We note that while we have made efforts to ensure that the API described here is as close as possible to the final implemented API, the authoritative version of the API will be given by the base classes that live in the core astropy package.'

Agree about not backporting.

**Comment by nden** [comment]
By default GWCS sets out-of-bounds to NaN. However, it's very useful in certain cases to be able to evaluate the transforms  without limits. I think it's a shortcoming of APE 14 that it doesn't allow passing `kwargs`. I understand why this is so but perhaps certain standard keywords can be allowed and an application can set them.

Another difference with GWCS is that while `pixel_bounds` don't have a setter, it is possible to reset them to None because they are tied to a property `bounding_box` which defines the bounds. Setting the `bounding_box` to None essentially sets `pixel_bounds` to None. Being able to reset the bounds allows flexibility which is perhaps not necessary for applications but quite useful for interactive work.

astropy.modeling expects numerical intervals for all directions, we'll have to add an option of `None` at some point to match this behaviour, and that's fine.

The failing tests are all in SlicedWCS, that's not too bad and is perhaps to be expected.
And finally, since I've been advocating for allowing this behavior from the start, I'd consider this a bug, not a new feature :-). But I agree with whatever way is easier/best to handle it.

**Comment by astrofrog** [comment]
I've now fixed the tests - the main failing tests were those for SlicedLowLevelWCS which had unrealistic pixel_bounds which didn't even include the CRPIX values. I've now updated this.

In practice, this should not impact many users because they would have needed to explicitly set ``pixel_bounds`` on a FITS WCS - it is never populated on reading from a FITS file.

**Comment by astrofrog** [comment]
Even though it's technically a bug, as it's an API change I think we can just go with 7.0.0

**Comment by pllim** [comment]
I think it needs a rebase to pick up updated test matrix.

**Comment by astrofrog** [comment]
Rebased!

**Comment by Cadair** [comment]
Just to be clear, the change is that if a user sets `WCS.pixel_bounds` then it will return NaN outside of them? There isn't a way to specify those bounds in a FITS header?

**Comment by astrofrog** [comment]
@Cadair - that's correct, we don't interpret anything from FITS headers as indicating any bounds so this will only affect setting bounds explicitly on the FITS ``WCS`` class.

File: astropy/wcs/wcsapi/tests/test_fitswcs.py [hunk 1]
```diff
@@ -1417,3 +1417,87 @@ def test_pixel_to_world_stokes(wcs_polarized):
...
+        else wcs.pixel_to_world_values
+    )
+
+    xp = np.arange(5) + 1
+    yp = np.arange(5) + 1
+
+    # Before setting bounds
+
+    # Scalars
+    xw, yw = func(xp[0], yp[0])
```
Comment (nden) [inline astropy/wcs/wcsapi/tests/test_fitswcs.py:1443]: `xp[0]` returns a type `np.int64`. Can you add a test with Python scalars?

Comment (astrofrog) [reply astropy/wcs/wcsapi/tests/test_fitswcs.py:1447]: Good catch, there was a bug with Python scalars! Fixed now.

**Comment by astrofrog** [comment]
I've now rebased this and fixed the case for Python scalars.

**Review by Cadair — APPROVED** [review]
lgtm

**Review by nden — APPROVED** [review]
LGTM

**Pull request merged**